lib/Runtime/Base/ObjectTypeCache.cpp

```diff
@@ -0,0 +1,51 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+#include "RuntimeBasePch.h"
+#include "Types/PathTypeHandler.h"
+
+namespace Js
+{
+
+DynamicType* ObjectTypeCache::FindOrCreateType(_In_ RecyclableObject* prototype)
+{
+    // For fast lookup, cache index is a function of prototype address
+    const uint index = (uint)(((uintptr_t)prototype) >> PolymorphicInlineCacheShift) & (ObjectTypeCache::MaxCachedTypes - 1);
+
+    RecyclerWeakReference<RecyclableObject>* cachedWeakProto = cache[index].prototype;
+
+    // Check for nullptr, because the cache will not have a weak ref yet if this is first access
+    if (cachedWeakProto != nullptr)
+    {
+        Assert(cache[index].type != nullptr);
+
+        DynamicType* cachedType = cache[index].type->Get();
+        RecyclableObject* cachedProto = cachedWeakProto->Get();
+
+        // We can use cache if the prototypes match and the Type hasn't been collected
+        if (cachedProto == prototype && cachedType != nullptr)
+        {
+            return cachedType;
+        }
+    }
+
+    // If we didn't already have a Type in our cache, we need to create a new one
+
+    JavascriptLibrary * library = prototype->GetLibrary();
+    ScriptContext * scriptContext = library->GetScriptContext();
+
+    SimplePathTypeHandler* typeHandler = SimplePathTypeHandler::New(scriptContext, library->GetRootPath(), 0, 0, 0, true, true);
+    DynamicType* newType = DynamicType::New(scriptContext, TypeIds_Object, prototype, nullptr, typeHandler, true, true);
+
+    // Store the Type and prototype as weak references to avoid unnecessarily extending their lifetimes
+    RecyclerWeakReference<RecyclableObject> *  newWeakProtoHandle = prototype->GetRecycler()->CreateWeakReferenceHandle(prototype);
```

Comment:
Considering how small this cache is, it's kind of nuts that we have to allocate 2 weak references (i.e 64 bytes) for every proto/pointer pair. I wonder if we can do something similar to inline caches here where we have a post mark cleanup step which simply scans this memory interpreting every pointer as a GC pointer and simply nulls it out if the item in question is not marked (and then have this cache be allocated as leaf instead)? Don't need to fix for this PR but worth thinking about

cc @atulkatti @jianchun